### PR TITLE
Use nvmlDeviceGetCount_v2() first for CUDA check

### PR DESCRIPTION
**Pull request by qkoziol**
Check for CUDA devices with nvmlDeviceGetCount_v2() first, to avoid more expensive call to cudaGetDeviceCount() when possible.

### CODE REVIEW

File: src/hmem_cuda.c
Comment (shefty) [inline src/hmem_cuda.c:480]: Declare variables at the top of the function.  It makes them easier to find.

File: configure.ac
Comment (shefty) [inline configure.ac:598]: What is the impact of requiring this new library?  Is it always available?

There's a missing set of checks somewhere, since we now need nvml.h present.

Comment (qkoziol) [reply configure.ac:598]: Yes, the nvml header is installed with CUDA, since 2012.  I don't think we need another check for it.

File: src/hmem_cuda.c
Comment (qkoziol) [reply src/hmem_cuda.c:434]: Yes, this should be installed with the CUDA libraries, and it's a failure if the install is wrong.

File: src/hmem_cuda.c
Comment (qkoziol) [reply src/hmem_cuda.c:474]: OK, will update

File: src/hmem_cuda.c
Comment (qkoziol) [reply src/hmem_cuda.c:480]: OK, will update

File: src/hmem_cuda.c
Comment (shijin-aws) [reply src/hmem_cuda.c:434]: ok, thanks.

**Comment by shijin-aws** [comment]
Please look at the AWS CI failure

File: src/hmem_cuda.c
Comment (shefty) [inline src/hmem_cuda.c:508]: Don't do updates in a separate patch.  Just update the original patch directly.  Otherwise, there are potentially 2 patches that need to be picked into the stable branches.

For the original patch, please format the commit messages, so that the 1st line is only about 72 characters long.  It needs to be easily viewable using git log --oneline for fast identification and cherry picking of patches.  Add more details into the main message.  This seems like a very minor optimization for a function that's called once.  Some details on why we're adding a second get device count call into this path should be captured in the commit message.

Comment (qkoziol) [reply src/hmem_cuda.c:508]: > Don't do updates in a separate patch. Just update the original patch directly. Otherwise, there are potentially 2 patches that need to be picked into the stable branches.
> 

Why not squash during the merge from my branch into main?

Comment (qkoziol) [reply src/hmem_cuda.c:540]: >  This seems like a very minor optimization for a function that's called once. Some details on why we're adding a second get device count call into this path should be captured in the commit message.

nvmlDeviceGetCount is much lighter-weight than cudaGetDeviceCount and allows an early-out from the routine, improving performance for a simple simple IMB Allreduce test on AWS instances without GPUs by 40-300%.

**Comment by qkoziol** [comment]
bot:aws:retest

File: src/hmem_cuda.c
Comment (shijin-aws) [reply src/hmem_cuda.c:540]: >Why not squash during the merge from my branch into main?

I don't quite follow. It is a convention in PRs that updates on a commit should be combined with that original commits via `git commit --amend`. There is no auto squash for github merge AFAIK

Comment (qkoziol) [reply src/hmem_cuda.c:540]: > > Why not squash during the merge from my branch into main?
> 
> I don't quite follow. It is a convention in PRs that updates on a commit should be combined with that original commits via `git commit --amend`. There is no such kind of "auto squash" for github merge AFAIK

All of the projects i've contributed to have done this in different ways.  ¯\_(ツ)_/¯

**Comment by qkoziol** [comment]
I'll close this PR and address all the formatting and commit-related issues by submitting another PR with a single commit covering the combined set of changes.

**Comment by shefty** [comment]
Please don't close the PR.  That loses the comments.  Just update the original patch and force push.

**Comment by qkoziol** [comment]
Squashed, updated the commit message, and force-pushed

**Review by shefty — APPROVED** [review]

**Review by shijin-aws — APPROVED** [review]

**Comment by shijin-aws** [comment]
We are evaluating the updated version before merging it.

**Comment by qkoziol** [comment]
Performance is still good, merging now.

**Comment by qkoziol** [comment]
@shefty - Should this be backported to any release branches?

**Comment by shefty** [comment]
I'll let AWS decide that.  IMO, it doesn't seem critical enough to backport.  I doubt many apps would notice this outside of some benchmarks.

**Pull request merged**
